sql/core/src/test/scala/org/apache/spark/sql/DataFrameAggregateSuite.scala

```diff
@@ -1545,6 +1545,158 @@ class DataFrameAggregateSuite extends QueryTest
+        hllsketch_binary("value").as("hllsketch_3")
+      )
+    df3.createOrReplaceTempView("df3")
+
+    // now test hllsketch_union_estimate via dataframe + sql, with and without configs,
+    // unioning together sketches with default, non-default and different configurations
+    val df4 = sql(
+      """select
+        | id,
+        | count(value),
+        | hllsketch_binary(value) as hllsketch_1,
+        | hllsketch_binary(value, 20, 'HLL_8') as hllsketch_2,
+        | hllsketch_binary(value, 20, 'HLL_8') as hllsketch_3
+        |from df2
+        |group by 1
+        |""".stripMargin)
+    df4.createOrReplaceTempView("df4")
+
+    val res3 = df3.union(df4).groupBy("id")
+      .agg(
+        sum("count").as("count"),
+        hllsketch_union_estimate("hllsketch_1", 20).as("distinct_count_1"),
+        hllsketch_union_estimate("hllsketch_2").as("distinct_count_2"),
+        hllsketch_union_estimate("hllsketch_3").as("distinct_count_3")
+      )
+    checkAnswer(res3, Row(1, 15, 6, 6, 6))
+
+    val res4 = sql(
+      """select
+        | id,
+        | sum(count) as count,
+        | hllsketch_union_estimate(hllsketch_1, 20) as distinct_count_1,
+        | hllsketch_union_estimate(hllsketch_2) as distinct_count_2,
+        | hllsketch_union_estimate(hllsketch_3) as distinct_count_3
+        |from (select * from df3 union all select * from df4)
+        |group by 1
+        |""".stripMargin)
+    checkAnswer(res4, Row(1, 15, 6, 6, 6))
+  }
+
+  test("SPARK-16484: hllsketch_estimate negative tests") {
+
+    val df1 = Seq(
+      (1, "a"), (1, "a"), (1, "a"),
+      (1, "b"),
+      (1, "c"), (1, "c"),
+      (1, "d")
+    ).toDF("id", "value")
+
+    // validate that the functions error out when lgConfigK < 0
+    val error0 = intercept[AnalysisException] {
+      val res = df1.groupBy("id")
+        .agg(
+          hllsketch_estimate("value", -1, "HLL_4").as("hllsketch")
+        )
+      checkAnswer(res, Nil)
+    }
+    assert(error0.toString contains "DATATYPE_MISMATCH")
+
+    val error1 = intercept[AnalysisException] {
+      val res = df1.groupBy("id")
+        .agg(
+          hllsketch_binary("value", -1, "HLL_4").as("hllsketch")
+        )
+      checkAnswer(res, Nil)
+    }
+    assert(error1.toString contains "DATATYPE_MISMATCH")
+
+    val error2 = intercept[AnalysisException] {
+      val res = df1.groupBy("id")
+        .agg(
+          hllsketch_binary("value").as("hllsketch")
+        )
+        .agg(
+          hllsketch_union_estimate("hllsketch", -1)
+        )
+      checkAnswer(res, Nil)
+    }
+    assert(error2.toString contains "DATATYPE_MISMATCH")
+
+    // validate that the functions error out with unsupported tgtHllType
+    val error3 = intercept[SparkException] {
+      val res = df1.groupBy("id")
+        .agg(
+          hllsketch_estimate("value", 12, "HLL_5").as("hllsketch")
+        )
+      checkAnswer(res, Nil)
+    }
+    assert(error3.toString contains "IllegalArgumentException")
```

Comment:
This is also not such a nice error to have, and I believe it does not follow the structure of error messages (does not have an associated error class...). Let's discuss about this.

Although this is not the right place to discuss this, I am wondering whether we really need to expose the `tgtHllType` as part of this PR. Also something to discuss.

Reply:
I think this is coming from Enum.valueOf. I could wrap the exception with a SketchesArgumentException (what gets thrown when instantiating an HllSketch with an invalid lgConfigK value)?

Reply:
I've removed the explicit checks in the aggregate function and instead am relying on the HllSketch/Union arg checks/exceptions.